Containers/docker-socket-proxy/haproxy.cfg

```diff
@@ -1,72 +1,24 @@
-    http-request allow if { path,url_dec -m reg -i ^(/v[\d\.]+)?/commit } { env(COMMIT) -m bool }
-    http-request allow if { path,url_dec -m reg -i ^(/v[\d\.]+)?/configs } { env(CONFIGS) -m bool }
-    http-request allow if { path,url_dec -m reg -i ^(/v[\d\.]+)?/containers } { env(CONTAINERS) -m bool }
-    http-request allow if { path,url_dec -m reg -i ^(/v[\d\.]+)?/distribution } { env(DISTRIBUTION) -m bool }
-    http-request allow if { path,url_dec -m reg -i ^(/v[\d\.]+)?/events } { env(EVENTS) -m bool }
-    http-request allow if { path,url_dec -m reg -i ^(/v[\d\.]+)?/exec } { env(EXEC) -m bool }
-    http-request allow if { path,url_dec -m reg -i ^(/v[\d\.]+)?/grpc } { env(GRPC) -m bool }
-    http-request allow if { path,url_dec -m reg -i ^(/v[\d\.]+)?/images } { env(IMAGES) -m bool }
-    http-request allow if { path,url_dec -m reg -i ^(/v[\d\.]+)?/info } { env(INFO) -m bool }
-    http-request allow if { path,url_dec -m reg -i ^(/v[\d\.]+)?/networks } { env(NETWORKS) -m bool }
-    http-request allow if { path,url_dec -m reg -i ^(/v[\d\.]+)?/nodes } { env(NODES) -m bool }
-    http-request allow if { path,url_dec -m reg -i ^(/v[\d\.]+)?/_ping } { env(PING) -m bool }
-    http-request allow if { path,url_dec -m reg -i ^(/v[\d\.]+)?/plugins } { env(PLUGINS) -m bool }
-    http-request allow if { path,url_dec -m reg -i ^(/v[\d\.]+)?/secrets } { env(SECRETS) -m bool }
-    http-request allow if { path,url_dec -m reg -i ^(/v[\d\.]+)?/services } { env(SERVICES) -m bool }
-    http-request allow if { path,url_dec -m reg -i ^(/v[\d\.]+)?/session } { env(SESSION) -m bool }
-    http-request allow if { path,url_dec -m reg -i ^(/v[\d\.]+)?/swarm } { env(SWARM) -m bool }
-    http-request allow if { path,url_dec -m reg -i ^(/v[\d\.]+)?/system } { env(SYSTEM) -m bool }
-    http-request allow if { path,url_dec -m reg -i ^(/v[\d\.]+)?/tasks } { env(TASKS) -m bool }
-    http-request allow if { path,url_dec -m reg -i ^(/v[\d\.]+)?/version } { env(VERSION) -m bool }
-    http-request allow if { path,url_dec -m reg -i ^(/v[\d\.]+)?/volumes } { env(VOLUMES) -m bool }
+    http-request deny unless { src 127.0.0.1 } || { src ::1 } || { src NC_IPV4_PLACEHOLDER } || { src NC_IPV6_PLACEHOLDER }
+    http-request allow if { path,url_dec -m reg -i ^(/v[\d\.]+)?/containers/[a-zA-Z0-9_.-]+/((start)|(stop)|(restart)|(kill)) }
+    http-request allow if { path,url_dec -m reg -i ^(/v[\d\.]+)?/auth }
+    http-request allow if { path,url_dec -m reg -i ^(/v[\d\.]+)?/containers }
+    http-request allow if { path,url_dec -m reg -i ^(/v[\d\.]+)?/images }
+    http-request allow if { path,url_dec -m reg -i ^(/v[\d\.]+)?/info }
+    http-request allow if { path,url_dec -m reg -i ^(/v[\d\.]+)?/networks }
+    http-request allow if { path,url_dec -m reg -i ^(/v[\d\.]+)?/_ping }
+    http-request allow if { path,url_dec -m reg -i ^(/v[\d\.]+)?/volumes }
+    http-request allow if { path,url_dec -m reg -i ^(/v[\d\.]+)?/version }
```

Comment:
As this docker_docker_proxy required only for app_ecosystem_v2, for now we need only these routes explicitly:

    http-request allow if { path,url_dec -m reg -i ^(/v[\d\.]+)?/containers/nc_app_[a-zA-Z0-9_.-]+/((json)|(start)|(stop)) } METH_GET
    http-request allow if { path,url_dec -m reg -i ^(/v[\d\.]+)?/containers/create?name=nc_app_[a-zA-Z0-9_.-]+ } METH_POST
    acl nc_app_volume_data req.body -m reg -i "\"name\":\s?\"nc_app_[a-zA-Z0-9_.-]+_data\""
    http-request allow if { path,url_dec -m reg -i ^(/v[\d\.]+)?/volumes/create } nc_app_volume_data METH_POST
    http-request allow if { path,url_dec -m reg -i ^(/v[\d\.]+)?/volumes/nc_app_[a-zA-Z0-9_.-]+_data } METH_DELETE
    http-request allow if { path,url_dec -m reg -i ^(/v[\d\.]+)?/images/create } METH_POST
    
With some restrictions of access to containers by name, and the same for volumes.

Reply:
Thank you Andrey! Do you think it would be possible to also disable bind-mounts for containers and also disable the privileged flag? I would like to make sure that no container mounts the docker socket itself nor runs in privileged mode. Would that be possible somehow? :)

Reply:
I think we could add additional ACLs to secure these requests.

```
...
    # ACL to restrict container creation, that it has HostConfig.Privileged only set to false
    acl no_privileged_flag req.body -m reg -i "\"HostConfig\":\s?{[^}]*\"Privileged\":\s?false"

    # ACL to restrict the number of Mounts to 1
    acl only_one_volume req.body -m reg -i "\"Mounts\":\s*\[\s*(?:(?!\"Mounts\":\s*\[)[^}]*)}[^}]*\]"

    # ACL to restrict the type of Mounts to volume (inverted match)
    acl type_volume_only req.body -m reg -i "\"Mounts\":\s?\[\s?{[^}]*\"Type\":\s?\"volume\""

    # ACL to allow only one volume mounted to container and that it is named nc_app_[a-zA-Z0-9_.-]+_data
    acl nc_app_volume_data_only req.body -m reg -i "\"Mounts\":\s?\[\s?{[^}]*\"Source\":\s?\"nc_app_[a-zA-Z0-9_.-]+_data\""

    http-request allow if { path,url_dec -m reg -i ^(/v[\d\.]+)?/containers/create?name=nc_app_[a-zA-Z0-9_.-]+ } no_privileged_flag only_one_volume type_volume_only nc_app_volume_data_only METH_POST
...
```

Correct me if I missed something.

Reply:
Thank you Andrey! 

> # ACL to restrict container creation, that it has HostConfig.Privileged only set to false
>     acl no_privileged_flag req.body -m reg -i "\"HostConfig\":\s?{[^}]*\"Privileged\":\s?false"

Regarding that regex I would probably simply test if Privileged is set at all because it should not be to fall back to the default of privileged: false - so simply do not allow privileged to be present in the hostconfig. WDYT?

---

> # ACL to restrict the number of Mounts to 1
>     acl only_one_volume req.body -m reg -i "\"Mounts\":\s*\[\s*(?:(?!\"Mounts\":\s*\[)[^}]*)}[^}]*\]"
> 
>     # ACL to restrict the type of Mounts to volume (inverted match)
>     acl type_volume_only req.body -m reg -i "\"Mounts\":\s?\[\s?{[^}]*\"Type\":\s?\"volume\""
> 
>     # ACL to allow only one volume mounted to container and that it is named nc_app_[a-zA-Z0-9_.-]+_data
>     acl nc_app_volume_data_only req.body -m reg -i "\"Mounts\":\s?\[\s?{[^}]*\"Source\":\s?\"nc_app_[a-zA-Z0-9_.-]+_data\""

This looks good but there is an additional way to set mounts via ['HostConfig']['Binds'] which should also be prevented imho.

---

Additionally we need to prevent the volume creation of specifying a host path. Would that be possible? :)

Reply:
> This looks good but there is an additional way to set mounts via ['HostConfig']['Binds'] which should also be prevented imho. - maybe we just disable setting ['HostConfig']['Binds'] at all?

acl no_binds req.body -m reg -I "\"HostConfig\":.*\"Binds\":"

Better to forbid using < **"Binds":** > string at all.

Reply:
Finally our restricting rules should look like that:

```http-request allow if { path,url_dec -m reg -i ^(/v[\d\.]+)?/containers/nc_app_[a-zA-Z0-9_.-]+/((json)|(start)|(stop)) } METH_GET
    http-request allow if { path,url_dec -m reg -i ^(/v[\d\.]+)?/containers/nc_app_[a-zA-Z0-9_.-]+ } METH_DELETE

    # ACL to restrict container name to nc_app_[a-zA-Z0-9_.-]+
    acl nc_app_container_name url_param(name) -m reg -i "^nc_app_[a-zA-Z0-9_.-]+"

    # ACL to restrict the number of Mounts to 1
    acl one_mount_volume req.body -m reg -i "\"Mounts\"\s*:\s*\[\s*(?:(?!\"Mounts\"\s*:\s*\[)[^}]*)}[^}]*\]"
    # ACL to deny if there are any binds
    acl binds_present req.body -m reg -i "\"HostConfig\"\s*:.*\"Binds\"\s*:"
    # ACL to restrict the type of Mounts to volume
    acl type_not_volume req.body -m reg -i "\"Mounts\":\s*\[[^\]]*(\"Type\":\s*\"(?!volume\b)\w+\"[^\]]*)+\]"
    http-request deny if { path,url_dec -m reg -i ^(/v[\d\.]+)?/containers/create } nc_app_container_name !one_mount_volume binds_present type_not_volume METH_POST

    # ACL to restrict container creation, that it has HostConfig.Privileged only set to false
    acl no_privileged_flag req.body -m reg -i "\"HostConfig\":\s?{[^}]*\"Privileged\":\s?false"
    # ACL to allow mount volume with strict pattern for name: nc_app_[a-zA-Z0-9_.-]+_data
    acl nc_app_volume_data_only req.body -m reg -i "\"Mounts\":\s?\[\s?{[^}]*\"Source\":\s?\"nc_app_[a-zA-Z0-9_.-]+_data\""
    http-request allow if { path,url_dec -m reg -i ^(/v[\d\.]+)?/containers/create } nc_app_container_name no_privileged_flag nc_app_volume_data_only METH_POST

    acl nc_app_volume_data req.body -m reg -i "\"Name\":\s?\"nc_app_[a-zA-Z0-9_.-]+_data\""
    http-request allow if { path,url_dec -m reg -i ^(/v[\d\.]+)?/volumes/create } nc_app_volume_data METH_POST
    http-request allow if { path,url_dec -m reg -i ^(/v[\d\.]+)?/volumes/nc_app_[a-zA-Z0-9_.-]+_data } METH_DELETE
    http-request allow if { path,url_dec -m reg -i ^(/v[\d\.]+)?/images/create } METH_POST```

Reply:
Thank you @andrey18106 ! :)

Does https://github.com/nextcloud/all-in-one/pull/3243/commits/b4e30bc66a56290d3f99906cacc93e49875b45c4 look good to you or did I remove too much or to less? :)